### ISSUE TITLE
cogen3: VARS

### ISSUE THREAD

**Issue by climbcat** [McStasMcXtrace/McCode#782]
Definitions:

- A particle user variable, or just user variable, is a parameter that is defined in the instrument VARS section, or though a CG macro in declare, which the cogen assigns as a particle state parameter.

- A symbolic parameter, or sympar, is a component parameter (previously a defpar), which is used as an alias for a user variable.

1) VARS variable and type extraction:
-Disallow comments in prototype
-Use a simple TYPE VARNAME SEMICOLON grammar
-Ensure name uniqueness
-Print extracted user variables to users (and intended access in trace)
-Comments should be trimmed (eventually)
-Apply the type as well as the varname in particle structs (eventually)

2) Sympar implementation:
-Allow sympar use in the grammar
-Add a def/undef around component trace functions, when the value of a sympar matches a user variable (Then user1 becomes synonymous with MyVar1, which was previously def'd as particle->MyVar1, kudos to @willend for coming up with this trick)
-Fail mcstas if the symbol value is unknown by the cogen as a user var
-Revert sympar parameters in Monitor_nD to use pure symbols rather than strings

3) Translate "user vars equivalent" using legacy instruments into using the above VARS implementation (such as ILL_BRISP, ESS_Brilliance_2001, etc.).

4) Elliminate any remaining legacy sympar using components to this format.

**Comment by climbcat** [comment]
One mystery solved: It turns out that component DECLARE code is copy-pasted verbatim into the component struct. There is no variable extraction. There is a check for "=", which produces a user friendly warning. (Search in cogen.c.in for "DECLARE code stored as structure members".)

The assumption is that anything "weird" would give rise to an (equally weird) compile error. There should be no reason to implement something more complicated when it comes to VARS...that section is copy-pasted into the particle struct as-is, including types.

**Comment by farhi** [comment]
Absolutely true. I assumed, when writing that, that authors of components and instrument would write sensible DECLARE sections. As these should only contain declarations without assigments (after removing comments), detecting an '=' signe thereafter triggers such a warning indeed.

**Comment by climbcat** [comment]
The component level side of things should, as the last big, have been implemented by 41c8ac6064ac0eca256222b849ca7859e3fbaf9d. Only double is supported on the component level (user1 etc.).

Added your warning as well, now users are warned if they try to assign in the USERVARS block. 

Questions: 
1) USERVARS was chosen because the grammar did not like things like Vars in component code, as long as I tried to use VARS as the block name.
2) As it is, symbol pars are setting parameters of the type double, which are declared with the symbol keyword. Is there another word that would work better for this purpose? (Perhaps user, var, var_d, or what have you.)
